types/event_bus.go

```diff
@@ -67,63 +68,106 @@ func (b *EventBus) Publish(eventType string, eventData TMEventData) error {
 
 //--- block, tx, and vote events
 
-func (b *EventBus) PublishEventNewBlock(block EventDataNewBlock) error {
-	return b.Publish(EventNewBlock, TMEventData{block})
+func (b *EventBus) PublishEventNewBlock(event EventDataNewBlock) error {
+	return b.Publish(EventNewBlock, TMEventData{event})
 }
 
-func (b *EventBus) PublishEventNewBlockHeader(header EventDataNewBlockHeader) error {
-	return b.Publish(EventNewBlockHeader, TMEventData{header})
+func (b *EventBus) PublishEventNewBlockHeader(event EventDataNewBlockHeader) error {
+	return b.Publish(EventNewBlockHeader, TMEventData{event})
 }
 
-func (b *EventBus) PublishEventVote(vote EventDataVote) error {
-	return b.Publish(EventVote, TMEventData{vote})
+func (b *EventBus) PublishEventVote(event EventDataVote) error {
+	return b.Publish(EventVote, TMEventData{event})
 }
 
-func (b *EventBus) PublishEventTx(tx EventDataTx) error {
+// PublishEventTx publishes tx event with tags from Result. Note it will add
+// predefined tags (EventTypeKey, TxHashKey). Existing tags with the same names
+// will be overwritten.
+func (b *EventBus) PublishEventTx(event EventDataTx) error {
 	// no explicit deadline for publishing events
 	ctx := context.Background()
-	b.pubsub.PublishWithTags(ctx, TMEventData{tx}, map[string]interface{}{EventTypeKey: EventTx, TxHashKey: fmt.Sprintf("%X", tx.Tx.Hash())})
+
+	tags := make(map[string]interface{})
+
+	// validate and fill tags from tx result
+	for _, tag := range event.Result.Tags {
+		// basic validation
+		if tag.Key == "" {
+			b.Logger.Info("Got tag with an empty key (skipping)", "tag", tag, "tx", event.Tx)
+			continue
+		}
+
+		switch tag.ValueType {
+		case abci.KVPair_STRING:
+			tags[tag.Key] = tag.ValueString
+		case abci.KVPair_INT:
+			tags[tag.Key] = tag.ValueInt
+		}
+	}
+
+	// add predefined tags
+	if tag, ok := tags[EventTypeKey]; ok {
```

Comment:
lets make this 3-liner a function, and include the key/value in the error statement